Allow overriding the "pretty" name of the OS inside the image

tjkirch opened the pull request:
**Issue number:**

Fixes #1314 - thanks @diranged!

**Description of changes:**

Similar to #1312, this passes another field through for user customization - this time the "pretty name" that shows up in os-release, which in turn is used by k8s and other tools.

The second commit reorganizes buildsys's builder module to use argument vectors instead of strings, so we can handle arbitrary characters, rather than using "|" as a marker (and subsequently making it unavailable to names).

**Testing done:**

Specifying a pretty name works:
```
> cargo make -e BUILDSYS_PRETTY_NAME="My Pretty Name" && cargo make ami -e BUILDSYS_PRETTY_NAME="My Pretty Name"
...
bash-5.0# cat /etc/os-release
PRETTY_NAME="My Pretty Name 1.0.5"
...
```
Default is still OK:
```
> cargo make && cargo make ami
...
bash-5.0# cat /etc/os-release 
PRETTY_NAME="Bottlerocket OS 1.0.5"
...
```

Retested with the second (refactor) commit and found that buildsys still produced the artifacts in the same way, and the resulting AMIs were still happy and healthy.  Here's one with a weird name:

```
> cargo make -e BUILDSYS_PRETTY_NAME="My|Pretty@Name" && cargo make ami -e BUILDSYS_PRETTY_NAME="My|Pretty@Name"
...
bash-5.0# cat /etc/os-release 
PRETTY_NAME="My|Pretty@Name 1.0.5"
...
bash-5.0# apiclient -u /os      
{"pretty_name":"My|Pretty@Name 1.0.5","variant_id":"aws-k8s-1.17","version_id":"1.0.5","build_id":"ace13fe9","arch":"x86_64"}
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

tjkirch committed:
Allow overriding the "pretty" name of the OS inside the image

samuelkarp commented on tools/buildsys/src/builder.rs:
```diff
@@ -93,6 +93,9 @@ impl VariantBuilder {
         let output_dir: PathBuf = getenv("BUILDSYS_OUTPUT_DIR")?.into();
+        // We expect users' PRETTY_NAME values to contain spaces for things like "Bottlerocket OS"
+        // and so we need to transform them the same way as PACKAGES above.
+        let pretty_name = getenv("BUILDSYS_PRETTY_NAME")?.replace(' ', "|");
```
I assume this means that `|` ends up being a reserved character that you can't use in the pretty name.  Is that correct?  Should we document that in the readme?

tjkirch replied:
A pipe would become a space. I've never seen that used before in an OS name. I think it'd be cluttering the docs... no?

samuelkarp replied:
Doesn't seem like it would be common, but it wouldn't surprise me to see someone want to use it.  I do think we should document it somewhere, but we could put it somewhere other than the readme if you think that's too much clutter.

tjkirch replied:
I'm going to think about alternatives in the code so this isn't required.

tjkirch replied:
Please see the new second commit and updated testing, I removed this character restriction.

zmrow reviewed (approved):
:cake:

tjkirch committed:
buildsys: use argument vecs instead of string manipulation

Some arguments have internal spaces, like the list of packages in --build-arg
PACKAGES=a b c.  This was handled using a "|" marker that was later replaced by
spaces.  For other uses of --build-arg, like PRETTY_NAME, we don't want to
arbitrarily restrict the characters that can be used, in case someone wants "|"
in their OS name.

This change stores arguments directly in vectors, rather than strings that we
split on space, so that we can pass through arbitrary characters.

tjkirch commented:
^ This push rebases on develop.

tjkirch commented:
> https://github.com/bottlerocket-os/bottlerocket/pull/1330/commits/ace13fe91d7c645ee1bc1d3abe18193f3d661f7e

The following push (which GitHub doesn't show distinctly) adds the above commit, which reorganizes the buildsys file to use argument vecs instead of strings, so that we can handle arbitrary characters.  This is intended to address @samuelkarp's concern.
